### PR TITLE
Consuming override in properties

### PR DESCRIPTION
**Pull request by kanika1894**
This will also handle the right click scenario.
Sample :
```
"properties": {     "ms.client.vscode": {
                            "overrides.inputMode": "{{#equals}} {{{client.rightClickScenario}}} true true 0 80{{/equals}}"}
                    }
```

### CODE REVIEW

File: src/configure/configure.ts [hunk 1]
```diff
@@ -186,8 +188,9 @@ class Orchestrator {
                 let extendedPipelineTemplate = await templateHelper.getTemplateParameteres(this.inputs.azureSession, (this.inputs.pipelineConfiguration.template as RemotePipelineTemplate));
                 let context: { [key: string]: any } = {};
                 context['subscriptionId'] = this.inputs.subscriptionId;
+                context['rightClickScenario'] = rightClickScenario;
```
Comment (anuragc617) [inline src/configure/configure.ts:191]: create it as private member and set the context values, wherever theses fields are set. Like subcriptionId at the place where subscriptionId is fetched.

File: src/configure/helper/mustacheHelper.ts [hunk 2]
```diff
@@ -3,11 +3,42 @@ import * as Mustache from 'mustache';
...
     public static getHelperMethods(): any {
         return {
-            "if": function () {
-                 /*
-                 * if returns first parameter if given clause is positive, otherwise second parameter
-                 * Usage: {{#if}}clause trueValue falseValue{{/if}}
+
+            "equals": function () {
+                /*
+                 * Usage: {{#equals}}value1 value2 true/false(ignorecase) returnIfTrue returnIfFalse(optional){{/regexReplace}}
```
Comment (anuragc617) [inline src/configure/helper/mustacheHelper.ts:9]: can you once verify the equals with Nitin and Tejas because I have seen different syntax in some templates..

File: src/configure/templateInputHelper/InputControl.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ export class InputControl {
     public dataSource: DataSourceExpression;
     public dataSourceInputControls: Array<InputControl>;
     public dataSourceInputs: Map<string, any>;
-    private inputDescriptor: ExtendedInputDescriptor;
+    public inputDescriptor: ExtendedInputDescriptor;
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControl.ts:15]: why this change?

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 4]
```diff
@@ -10,24 +10,41 @@ import { VisibilityHelper } from "./utilities/VisibilityHelper";
...
     constructor(pipelineTemplate: ExtendedPipelineTemplate, context: { [key: string]: any }) {
         this._pipelineTemplate = pipelineTemplate;
         this._inputControlsMap = new Map<string, InputControl>();
+        this._context = context;
         this._createControls(context);
     }
 
-    public async getAllPipelineTemplateInputs(azureSession: AzureSession, resourceNode?: GenericResource) {
+    public async overrideParameters(inputControl: InputControl) {
+        var properties = inputControl.getPropertyValue(constants.clientPropertyKey);
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:23]: use 'let' or 'const' instead of 'var' everywhere.

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 5]
```diff
@@ -10,24 +10,41 @@ import { VisibilityHelper } from "./utilities/VisibilityHelper";
...
 
     constructor(pipelineTemplate: ExtendedPipelineTemplate, context: { [key: string]: any }) {
         this._pipelineTemplate = pipelineTemplate;
         this._inputControlsMap = new Map<string, InputControl>();
+        this._context = context;
         this._createControls(context);
     }
 
-    public async getAllPipelineTemplateInputs(azureSession: AzureSession, resourceNode?: GenericResource) {
+    public async overrideParameters(inputControl: InputControl) {
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:22]: this function is not async remove it. Add void return type.

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 6]
```diff
@@ -10,24 +10,41 @@ import { VisibilityHelper } from "./utilities/VisibilityHelper";
...
+            var dependentClientInputMap = this._getClientDependencyMap(inputControl, [properties[element]]);
+            var newValue = this._computeMustacheValue(properties[element], dependentInputControlArray, dependentClientInputMap);
+            inputControl.inputDescriptor[key] = newValue;
+            if (key === "inputMode") {
+                inputControl.updateInputControlType(parseInt(newValue));
+            }
+        });
+    }
+
+    public async getAllPipelineTemplateInputs(azureSession: AzureSession) {
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:38]: also add return type for this function

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 7]
```diff
@@ -87,21 +102,22 @@ export class InputControlProvider {
...
 
-    private _computeMustacheValue(mustacheExpression: string, dependentInputControlArray: InputControl[]): string {
+    private _computeMustacheValue(mustacheExpression: string, dependentInputControlArray: InputControl[], dependentClientInputMap: StringMap<any>): string {
 
         var dependentInputValues = this._getInputParameterValueIfAllSet(dependentInputControlArray);
         if (dependentInputControlArray && dependentInputControlArray.length > 0 && !dependentInputValues) {
             return "";
         } else {
-            return MustacheHelper.render(mustacheExpression, { inputs: dependentInputValues });
+            return MustacheHelper.render(mustacheExpression, { inputs: dependentInputValues, client: dependentClientInputMap });
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:120]: can you please figure out what it needs to be present in system context as system context is also being used  in templates.

File: src/configure/templateInputHelper/InputControl.ts [hunk 8]
```diff
@@ -62,6 +62,24 @@ export class InputControl {
         this.visible = value;
     }
 
+    public updateInputControlType(inputMode: InputMode): void {
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControl.ts:65]: I think we can move the logic to update and intialize the inputMode at one place in InputControlUtility and use it here as well as in InputControl.createControls()

File: src/configure/helper/mustacheHelper.ts (hunk 2 above)
Comment (kanika1894) [reply src/configure/helper/mustacheHelper.ts:9]: Verified with both of them

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 5 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:22]: updated

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 6 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:38]: updated

File: src/configure/configure.ts (hunk 1 above)
Comment (kanika1894) [reply src/configure/configure.ts:191]: Updated

File: src/configure/templateInputHelper/InputControl.ts (hunk 3 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControl.ts:15]: As discussed, created separate function

File: src/configure/templateInputHelper/InputControl.ts (hunk 8 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControl.ts:65]: Updated

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 4 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:23]: done

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 7 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:120]: As discussed, will be handling later while working with yml files

File: src/configure/templateInputHelper/utilities/InputControlUtility.ts [hunk 9]
```diff
@@ -1,21 +1,44 @@
...
+            case InputMode.PasswordBox:
+                return ControlType.InputBox;
+            case InputMode.Combo:
+            case InputMode.CheckBox:
+            case InputMode.RadioButtons:
+                return ControlType.QuickPick;
+            default:
+                return null;
+        }
+    }
```
Comment (anuragc617) [inline src/configure/templateInputHelper/utilities/InputControlUtility.ts:41]: add new line

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 10]
```diff
@@ -168,7 +167,33 @@ export class InputControlProvider {
...
+        for (var clientInput of uniqueDependentClientInputs) {
+            var dependentInputControl = this._context[clientInput];
+            if (dependentInputControl) {
+                dependentClientControlMap[clientInput] = dependentInputControl;
+            } else {
+                throw new Error(`Dependent client input ${clientInput} specified is not present in client context.`);
+            }
+        }
+        return dependentClientControlMap;
+    }
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:196]: add new line after function

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 11]
```diff
@@ -168,7 +167,33 @@ export class InputControlProvider {
                 ? "" : inputDes.defaultValue;
         }
     }
+    private _getClientDependencyMap(inputControl: InputControl, dependencyExpressionArray: string[]) {
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:170]: new line in start of function

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 12]
```diff
@@ -51,110 +69,91 @@ export class InputControlProvider {
...
-                this._inputControlsMap.set(input.id, inputControl);
+            if (input.type !== InputDataType.Authorization) {
+                let controlType: ControlType = InputControlUtility.getInputControlType(input.inputMode);
+                inputControl = new InputControl(input, inputControlValue, controlType, this.azureSession);
+                if (inputControl) {
+                    this._inputControlsMap.set(input.id, inputControl);
+                    this._setInputControlDataSourceInputs(inputControl);
+                }
             }
+            this._initializeDynamicValidations(input);
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:80]: this should be called after overrideParameters in getAllPipelineTemplateInputs() and its argument should be inputControl.

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 13]
```diff
@@ -51,110 +69,91 @@ export class InputControlProvider {
...
-
-            }
-            if (inputControl) {
-                this._inputControlsMap.set(input.id, inputControl);
+            if (input.type !== InputDataType.Authorization) {
+                let controlType: ControlType = InputControlUtility.getInputControlType(input.inputMode);
+                inputControl = new InputControl(input, inputControlValue, controlType, this.azureSession);
+                if (inputControl) {
+                    this._inputControlsMap.set(input.id, inputControl);
+                    this._setInputControlDataSourceInputs(inputControl);
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:77]: this should aslo be called after overrideParameters in getAllPipelineTemplateInputs()

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 14]
```diff
@@ -28,14 +29,31 @@ export class InputControlProvider {
...
+
     public async getAllPipelineTemplateInputs() {
         let parameters: { [key: string]: any } = {};
         for (let inputControl of this._inputControlsMap.values()) {
             if (this.repoAnalysisSettingInputProvider.inputFromRepoAnalysisSetting(inputControl)) {
                 await this.repoAnalysisSettingInputProvider.setInputControlValueFromRepoAnalysisResult(inputControl);
             }
-            else if (inputControl.getPropertyValue('deployTarget') === "true" && !!this._context["resourceId"]) {
-                inputControl.setValue(this._context["resourceId"]);
+            else if (!!inputControl.getPropertyValue(constants.clientPropertyKey)) {
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:55]: logic here should be first override parameters of inputControl and then set its dataSourceControls, initializaDynamicValidations. After this set the value from repoAnalysis or else condition.
code in for loop should be like:
(!!inputControl.getPropertyValue(constants.clientPropertyKey)) {
   this.overrideParameters(inputControl);
}
this._setInputControlDataSourceInputs(inputControl);
this._initializeDynamicValidations(inputControl);
 if (this.repoAnalysisSettingInputProvider.inputFromRepoAnalysisSetting(inputControl)) {
                await this.repoAnalysisSettingInputProvider.setInputControlValueFromRepoAnalysisResult(inputControl);	                } else {
    this._setInputControlVisibility(inputControl);
    this._setupInputControlDefaultValue(inputControl);
     await inputControl.setInputControlValue();
}

File: src/configure/templateInputHelper/utilities/InputControlUtility.ts (hunk 9 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/utilities/InputControlUtility.ts:41]: added

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 10 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:196]: done

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 12 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:80]: updated

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 14 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:55]: Updated

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 13 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:77]: updated

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 15]
```diff
@@ -28,15 +29,34 @@ export class InputControlProvider {
         this._createControls();
     }
 
+    public async overrideParameters(inputControl: InputControl) {
```
Comment (anuragc617) [inline src/configure/templateInputHelper/InputControlProvider.ts:32]: make it private and move this function after public functions.
